src/GraphQL/Execution/ExecutionHelper.cs

```diff
@@ -72,195 +72,248 @@ public static FieldType GetFieldDefinition(ISchema schema, IObjectGraphType pare
+                    // attempt to retrieve the variable value from the inputs
+                    if (inputs.TryGetValue(variableDef.Name, out var variableValue))
                     {
-                        variable.Value = GetVariableValue(document, schema, v, variableValue);
+                        // parse the variable via ParseValue (for scalars) and ParseDictionary (for objects) as applicable
+                        variable.Value = GetVariableValue(document, graphType, variableDef, variableValue);
                     }
-                    else
+                    else if (variableDef.DefaultValue != null)
                     {
-                        var value = GetVariableValue(document, schema, v, v.DefaultValue?.Value);
-                        if (value != null)
-                            variable.Value = value;
+                        // if the variable was not specified in the inputs, and a default literal value was specified, use the specified default variable value
+
+                        // parse the variable literal via ParseLiteral (for scalars) and ParseDictionary (for objects) as applicable
+                        variable.Value = CoerceValue(schema, graphType, variableDef.DefaultValue, variables, null).Value;
+                        variable.IsDefault = true;
                     }
+                    else if (graphType is NonNullGraphType)
+                    {
+                        ThrowNullError(variable.Name);
+                    } 
+
+                    // if the variable was not specified and no default was specified, do not add the variable to the Variables object
 
+                    // add the variable to the list of parsed variables defined for the operation
                     variables.Add(variable);
                 }
             }
 
+            // return the list of parsed variables defined for the operation
             return variables;
         }
 
+        private static void ThrowNullError(string variableName)
+            => throw new InvalidVariableError(variableName, "Received a null input for a non-null variable.");
+
         /// <summary>
         /// Return the specified variable's value for the document from the attached <see cref="Inputs"/> object.
+        /// <br/><br/>
+        /// Validates and parses the supplied input object according to the variable's type, and converts the object
+        /// with <see cref="ScalarGraphType.ParseValue(object)"/> and
+        /// <see cref="IInputObjectGraphType.ParseDictionary(IDictionary{string, object})"/> as applicable.
+        /// <br/><br/>
         /// Since v3.3, returns null for variables set to null rather than the variable's default value.
         /// </summary>
-        public static object GetVariableValue(Document document, ISchema schema, VariableDefinition variable, object input)
+        public static object GetVariableValue(Document document, IGraphType graphType, VariableDefinition variable, object input)
```

Comment:
Do we need this method in public API?

Reply:
Probably not.  Do you think it should be made private?

Reply:
Yes. I tend to reduce the public API if we do not see an obvious need for it. This will make life easier for us.